TDDWithURMAndKotlin/part3_jump.md

```diff
@@ -0,0 +1,176 @@
+
+#### What?!?
+
+Relax, it's easier than you think. Let's see an example:
+
+You have the the registry:
+
+```none
+X: [,,4]
+```
+
+And the instruction:
+
+```none
+1: Z(2)
+2: J(2, 3, 10)
+3: I(2)
+4: J(2, 2, 2)
+```
+
+This set of instructions will only clone a number in a registry into another registry. In this case the registry at position 2 will end having the same value of registry 3. Don't believe me? No problem! Let's do a dry run seeing what happens in memory:
+
+```none
+1: Z(2)        -> X: [,0,4]
+2: J(2, 3, 10) -> X: [,0,4] // 0 != 4 so we continue execution at line 3
+3: I(2)        -> X: [,1,4]
+4: J(2, 2, 2)  -> X: [,1,4] // 1 == 1 comparing to the same position is always true, so we move back to line 2
+
+2: J(2, 3, 10) -> X: [,1,4] // 1 != 4 so we continue execution at line 3
+3: I(2)        -> X: [,2,4]
+4: J(2, 2, 2)  -> X: [,2,4] // 2 == 2 comparing to the same position is always true, so we move back to line 2
+
+2: J(2, 3, 10) -> X: [,2,4] // 2 != 4 so we continue execution at line 3
+3: I(2)        -> X: [,3,4]
+4: J(2, 2, 2)  -> X: [,3,4] // 3 == 3 comparing to the same position is always true, so we move back to line 2
+
+2: J(2, 3, 10) -> X: [,3,4] // 3 != 4 so we continue execution at line 3
+3: I(2)        -> X: [,4,4]
+4: J(2, 2, 2)  -> X: [,4,4] // 4 == 4 comparing to the same position is always true, so we move back to line 2
+
+2: J(2, 3, 10) -> X: [,3,4] // 4 == 4 so let's move to instruction 10… wait… there's no instruction 10, so we finish the execution
+```
+
+As you can see the `J` a.k.a `jump` function is a simple `go-to` in our machine. It moves the execution from one point to another. We need to create a way to represent this on our interpreter.
+
+## But first… tests
+
+To begin with our `jump` function we need a test for it. We know the function will take a registry to fetch the values to compare and an integer to move the execution to that line. for now, we will pass the set of instructions to the function to allow the `jump` function to move the _current_ index instruction.
+We can take advantage of our other functions already created to setup some registers and then check the function works as expected:
+
+```kotlin
+@Test
+fun `jump function should move to instruction X when registries are equal`() {
+    zero(registry, 5)
+    zero(registry, 10)
+
+    jump(registry, 5, 10, instructionSet, 1)
+
+    assertEquals(1, instructionSet.current)
+}
+```
+
+Wow, that's actually an ugly function. It takes 5 parameters, probably Uncle Bob is looking at us and feeling disappointed… well, we will refactor it later. For now let's see if it works… well, actually we know it doesn't work, as [we see previously](https://dev.to/sierisimo/tdd-with-urm-and-kotlin-1dj7), if we run `gradle check` to run the tests, it will generate a _compile-time error_ because we haven't created the `jump` function yet.
```

Comment:
```suggestion
Wow, that's actually an ugly function. It takes 5 parameters, probably Uncle Bob is looking at us and feeling disappointed… well, we will refactor it later. For now let's see if it works… well, actually we know it doesn't work, as [we have seen previously](https://dev.to/sierisimo/tdd-with-urm-and-kotlin-1dj7), if we run `gradle check` to run the tests, it will generate a _compile-time error_ because we haven't created the `jump` function yet.
```
The past is the past, even if it's not perfect, but this is a case for past perfect. NVM, I ran out of puns. For now.